modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/AbstractKeyValueStorage.java

```diff
@@ -79,22 +91,30 @@ public abstract class AbstractKeyValueStorage implements KeyValueStorage {
     protected final AtomicBoolean stopCompaction = new AtomicBoolean();
 
     /** Tracks only cursors, since reading a single entry or a batch is done entirely under {@link #rwLock}. */
-    protected final ReadOperationForCompactionTracker readOperationForCompactionTracker = new ReadOperationForCompactionTracker();
+    protected final ReadOperationForCompactionTracker readOperationForCompactionTracker;
 
-    /**
-     * Used to generate read operation ID for {@link #readOperationForCompactionTracker}.
-     *
-     * <p>Multi-threaded access is guarded by {@link #rwLock}.</p>
-     */
-    protected long readOperationIdGeneratorForTracker;
+    protected final ExecutorService compactionExecutor;
+
+    private final List<CompactionListener> compactionListeners = new CopyOnWriteArrayList<>();
```

Comment:
We already have the EventProducer interface, can it be used instead?

Reply:
So far I don't see any point in one event.